src/internal/colors/HtmlColor.cpp

```diff
@@ -36,28 +36,29 @@ static inline char hexdigit(uint8_t v)
 }
 
 
-size_t HtmlColor::ToNumericalString(char* buf, size_t bufSize) const
-{
-    size_t bufLen = bufSize - 1;
-
-    if (bufLen-- > 0)
-    {
-        if (bufLen > 0)
-        {
-            buf[0] = '#';
-        }
-
-        uint32_t color = Color;
-        for (uint8_t indexDigit = 6; indexDigit > 0; indexDigit--)
-        {
-            if (bufLen > indexDigit)
-            {
-                buf[indexDigit] = hexdigit(color & 0x0000000f);
-            }
-            color >>= 4;
-        }
-
-        buf[(bufLen < 7 ? bufLen : 7)] = 0;
+/**
+ * Generates a Html encoded hex color string (#aabbcc) with null termination.
+ *
+ * @param buf the buffer to write the string to
+ * @param bufSize the maximum buffer size (8 recommended for # + 123456 + null terminator)
+ * @return The amount of chars written to buf.
+ */
+size_t HtmlColor::ToNumericalString(char *buf, size_t bufSize) const {
+  if (bufSize > 0) {
+    buf[0] = '#';
+
+    uint32_t color = Color;
+    for (uint8_t indexDigit = 6; indexDigit > 0; --indexDigit) {
+      if (bufSize > indexDigit) {
+        buf[indexDigit] = hexdigit(color & 0x0000000f);
+      }
```

Comment:
checking something in the inner loop that can be calculated once, consider the following...
```
    size_t indexLast = min(7, bufSize - 1);
    buf[indexLast] = '\0';
    for (size_t indexDigit = indexLast; indexDigit > 0; --indexDigit) // note pre-decrement
    {
        buf[indexDigit] = hexdigit(color & 0x0000000f);
        color >>= 4;
    }
```

Reply:
The pre decrement is something I learned at work from someone I'd call a C/C++ expert. Never write for loops with post increment if you're not sure the compiler can handle this optimization. With ++i only the result is important in any case, i++ may not be optimized by all compilers, because the initial value might still be important when looking at that part.

https://embeddedgurus.com/stack-overflow/2008/08/efficient-c-tips-3-avoiding-post-increment-decrement/
https://stackoverflow.com/questions/24886/is-there-a-performance-difference-between-i-and-i-in-c

Reply:
Whenever there is a pre-operation, I just put a comment there as it's easy to miss and does make the code harder to quickly understand.  No need to remove the pre operations, just note them on a complex line.
Those are all good points about pre/post; but it's a minor optimization that most modern compilers do handle well.  But compared to the readability/maintainability of the code is the issue.  Note that we have bugs here ;-)